### PR TITLE
Pipelining of multiple file reads

### PR DESCRIPTION
**Pull request by lidavidm**
See this execution trace for performance issues.
https://gist.github.com/lidavidm/99aa79c9a5667409c006132e6162d387

### CODE REVIEW

File: cpp/parquetjni/parquetjni.cc [hunk 1]
```diff
@@ -229,46 +262,85 @@ class DatasetBatchReader : public arrow::RecordBatchReader {
...
+              return BufferedFile{status, nullptr, nullptr};
+            }
+            TRACE_WITH("ParquetJniDataset::EnqueueReads", next_file, profile_start,
+                       load_end);
+            return BufferedFile(arrow::Status::OK(),
+                                std::unique_ptr<ParquetJniReader>(jni_reader),
+                                std::move(batch_reader));
+          });
+      std::future<BufferedFile> future = task.get_future();
+      std::thread thread(std::move(task));
```
Comment (mayuropensource) [inline cpp/parquetjni/parquetjni.cc:327]: can we use threadpool so that we can reuse threads?

File: cpp/parquetjni/parquetjni.cc [hunk 2]
```diff
@@ -229,46 +262,85 @@ class DatasetBatchReader : public arrow::RecordBatchReader {
...
   }
 
   arrow::Status LoadNext() {
-    current_reader_ = nullptr;
-    if (current_jni_reader_) {
-      RETURN_NOT_OK(current_jni_reader_->Close());
+    if (!file_readers_.empty()) {
+      VLOG(2) << "Closing a file";
+      BufferedFile &buffered_file =
+          const_cast<BufferedFile &>(file_readers_.front().get());
```
Comment (mayuropensource) [inline cpp/parquetjni/parquetjni.cc:292]: is there a pop_front()?

File: cpp/parquetjni/parquetjni.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/parquetjni/parquetjni.cc:327]: C++ doesn't have a built in pool. I may convert this to use a Boost pool.

File: cpp/parquetjni/parquetjni.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/parquetjni/parquetjni.cc:292]: Unfortunately no. https://en.cppreference.com/w/cpp/container/queue

File: cpp/parquetjni/parquetjni.cc (hunk 1 above)
Comment (mayuropensource) [reply cpp/parquetjni/parquetjni.cc:327]: I see. That's fine then.

File: cpp/parquetjni/parquetjni.cc (hunk 2 above)
Comment (mayuropensource) [reply cpp/parquetjni/parquetjni.cc:292]: ok

**Review by mayuropensource — APPROVED** [review]

**Review by mayuropensource — APPROVED** [review]

**Pull request merged**